File: go/test/meeting_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/zoom-sdk/go"
	"github.com/voxgig-sdk/zoom-sdk/go/core"

	vs "github.com/voxgig-sdk/zoom-sdk/go/utility/struct"
)

func TestMeetingEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Meeting(nil)
		if ent == nil {
			t.Fatal("expected non-nil MeetingEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"meeting": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Meeting(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Meeting(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := meetingBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "meeting." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set ZOOM_TEST_MEETING_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		meetingRef01Ent := client.Meeting(nil)
		meetingRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "meeting"}, setup.data), "meeting_ref01"))
		meetingRef01Data["user_id"] = setup.idmap["user01"]

		meetingRef01DataResult, err := meetingRef01Ent.Create(meetingRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		meetingRef01Data = core.ToMapAny(entityData(meetingRef01DataResult))
		if meetingRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if meetingRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		meetingRef01Match := map[string]any{
			"user_id": setup.idmap["user01"],
		}

		meetingRef01ListResult, err := meetingRef01Ent.List(meetingRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		meetingRef01List, meetingRef01ListOk := meetingRef01ListResult.([]any)
		if !meetingRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", meetingRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(meetingRef01List), map[string]any{"id": meetingRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		meetingRef01DataUp0Up := map[string]any{
			"id": meetingRef01Data["id"],
		}

		meetingRef01MarkdefUp0Name := "agenda"
		meetingRef01MarkdefUp0Value := fmt.Sprintf("Mark01-meeting_ref01_%d", setup.now)
		meetingRef01DataUp0Up[meetingRef01MarkdefUp0Name] = meetingRef01MarkdefUp0Value

		meetingRef01ResdataUp0Result, err := meetingRef01Ent.Update(meetingRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		meetingRef01ResdataUp0 := core.ToMapAny(entityData(meetingRef01ResdataUp0Result))
		if meetingRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if meetingRef01ResdataUp0["id"] != meetingRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if meetingRef01ResdataUp0[meetingRef01MarkdefUp0Name] != meetingRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", meetingRef01MarkdefUp0Name, meetingRef01ResdataUp0[meetingRef01MarkdefUp0Name])
		}

		// LOAD
		meetingRef01MatchDt0 := map[string]any{
			"id": meetingRef01Data["id"],
		}
		meetingRef01DataDt0Loaded, err := meetingRef01Ent.Load(meetingRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		meetingRef01DataDt0LoadResult := core.ToMapAny(entityData(meetingRef01DataDt0Loaded))
		if meetingRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if meetingRef01DataDt0LoadResult["id"] != meetingRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		meetingRef01MatchRm0 := map[string]any{
			"id": meetingRef01Data["id"],
		}
		_, err = meetingRef01Ent.Remove(meetingRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		meetingRef01MatchRt0 := map[string]any{
			"user_id": setup.idmap["user01"],
		}

		meetingRef01ListRt0Result, err := meetingRef01Ent.List(meetingRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		meetingRef01ListRt0, meetingRef01ListRt0Ok := meetingRef01ListRt0Result.([]any)
		if !meetingRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", meetingRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(meetingRef01ListRt0), map[string]any{"id": meetingRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func meetingBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "meeting", "MeetingTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read meeting test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse meeting test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"meeting01", "meeting02", "meeting03", "user01", "user02", "user03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("ZOOM_TEST_MEETING_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"ZOOM_TEST_MEETING_ENTID": idmap,
		"ZOOM_TEST_LIVE":      "FALSE",
		"ZOOM_TEST_EXPLAIN":   "FALSE",
		"ZOOM_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["ZOOM_TEST_MEETING_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["ZOOM_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["ZOOM_APIKEY"],
			},
			extra,
		})
		client = sdk.NewZoomSDK(core.ToMapAny(mergedOpts))
	}

	live := env["ZOOM_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["ZOOM_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/meeting_entity_test.lua
-- Meeting entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("zoom_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("MeetingEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Meeting(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["meeting"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:Meeting(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:Meeting(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = meeting_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "meeting." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set ZOOM_TEST_MEETING_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local meeting_ref01_ent = client:Meeting(nil)
    local meeting_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.meeting"), "meeting_ref01"))
    meeting_ref01_data["user_id"] = setup.idmap["user01"]

    local meeting_ref01_data_result, err = meeting_ref01_ent:create(meeting_ref01_data, nil)
    assert.is_nil(err)
    meeting_ref01_data = helpers.to_map(type(meeting_ref01_data_result) == 'table' and meeting_ref01_data_result.data_get and meeting_ref01_data_result:data_get() or meeting_ref01_data_result)
    assert.is_not_nil(meeting_ref01_data)
    assert.is_not_nil(meeting_ref01_data["id"])

    -- LIST
    local meeting_ref01_match = {
      ["user_id"] = setup.idmap["user01"],
    }

    local meeting_ref01_list_result, err = meeting_ref01_ent:list(meeting_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(meeting_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(meeting_ref01_list_result),
      { id = meeting_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local meeting_ref01_data_up0_up = {
      id = meeting_ref01_data["id"],
    }

    local meeting_ref01_markdef_up0_name = "agenda"
    local meeting_ref01_markdef_up0_value = "Mark01-meeting_ref01_" .. tostring(setup.now)
    meeting_ref01_data_up0_up[meeting_ref01_markdef_up0_name] = meeting_ref01_markdef_up0_value

    local meeting_ref01_resdata_up0_result, err = meeting_ref01_ent:update(meeting_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local meeting_ref01_resdata_up0 = helpers.to_map(type(meeting_ref01_resdata_up0_result) == 'table' and meeting_ref01_resdata_up0_result.data_get and meeting_ref01_resdata_up0_result:data_get() or meeting_ref01_resdata_up0_result)
    assert.is_not_nil(meeting_ref01_resdata_up0)
    assert.are.equal(meeting_ref01_resdata_up0["id"], meeting_ref01_data_up0_up["id"])
    assert.are.equal(meeting_ref01_resdata_up0[meeting_ref01_markdef_up0_name], meeting_ref01_markdef_up0_value)

    -- LOAD
    local meeting_ref01_match_dt0 = {
      id = meeting_ref01_data["id"],
    }
    local meeting_ref01_data_dt0_loaded, err = meeting_ref01_ent:load(meeting_ref01_match_dt0, nil)
    assert.is_nil(err)
    local meeting_ref01_data_dt0_load_result = helpers.to_map(type(meeting_ref01_data_dt0_loaded) == 'table' and meeting_ref01_data_dt0_loaded.data_get and meeting_ref01_data_dt0_loaded:data_get() or meeting_ref01_data_dt0_loaded)
    assert.is_not_nil(meeting_ref01_data_dt0_load_result)
    assert.are.equal(meeting_ref01_data_dt0_load_result["id"], meeting_ref01_data["id"])

    -- REMOVE
    local meeting_ref01_match_rm0 = {
      id = meeting_ref01_data["id"],
    }
    local _, err = meeting_ref01_ent:remove(meeting_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local meeting_ref01_match_rt0 = {
      ["user_id"] = setup.idmap["user01"],
    }

    local meeting_ref01_list_rt0_result, err = meeting_ref01_ent:list(meeting_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(meeting_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(meeting_ref01_list_rt0_result),
      { id = meeting_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function meeting_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/meeting/MeetingTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read meeting test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "meeting01", "meeting02", "meeting03", "user01", "user02", "user03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("ZOOM_TEST_MEETING_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["ZOOM_TEST_MEETING_ENTID"] = idmap,
    ["ZOOM_TEST_LIVE"] = "FALSE",
    ["ZOOM_TEST_EXPLAIN"] = "FALSE",
    ["ZOOM_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["ZOOM_TEST_MEETING_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["ZOOM_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["ZOOM_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["ZOOM_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["ZOOM_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
